utils/src/main/java/com/cloud/utils/ssh/SSHCmdHelper.java

```diff
@@ -77,8 +80,47 @@ public String getStdErr() {
     }
 
     public static com.trilead.ssh2.Connection acquireAuthorizedConnection(String ip, int port, String username, String password) {
+        return acquireAuthorizedConnection(ip, 22, username, password, null);
+    }
+
+    public static boolean acquireAuthorizedConnectionWithPublicKey(final com.trilead.ssh2.Connection sshConnection, final String username, final String privateKey) {
+        if (StringUtils.isNotBlank(privateKey)) {
+            File privateKeyFile = null;
+            try {
+                privateKeyFile = File.createTempFile("cloudstack-host-", null);
+                FileWriter writer = new FileWriter(privateKeyFile.getAbsolutePath());
+                writer.write(privateKey);
+                writer.close();
```

Comment:
this is a bit of a security risk, it seems. The file is created in /tmp and never deleted (automatically if the JVM shuts down, but what if it crashes?). We may want a way to pass the key to the `sshConnection` as string instead of a file ref.

Reply:
@DaanHoogland 
the file is deleted in the `finally` block of next `try..catch..finally`

Reply:
@DaanHoogland 
you are right. we could use `char[]`  instead of `File` in the authentication. i will change it.